core-processor/src/executor.rs

```diff
@@ -47,17 +47,27 @@ fn make_checks_and_charge_gas_for_pages<'a>(
     allocations: &BTreeSet<WasmPageNumber>,
     pages_with_data: impl Iterator<Item = &'a PageNumber>,
     static_pages: WasmPageNumber,
+    initial_execution: bool,
 ) -> Result<WasmPageNumber, ExecutionErrorReason> {
     // Checks that all pages with data are in allocations set.
     for page in pages_with_data {
-        if !allocations.contains(&page.to_wasm_page()) {
+        let wasm_page = page.to_wasm_page();
+        if wasm_page >= static_pages && !allocations.contains(&wasm_page) {
             return Err(ExecutionErrorReason::PageIsNotAllocated(*page));
         }
     }
 
-    let mem_size = if let Some(max_wasm_page) = allocations.iter().next_back() {
+    let mem_size = if !initial_execution {
+        let max_wasm_page = if let Some(page) = allocations.iter().next_back() {
+            *page
+        } else if static_pages.0 != 0 {
```

Comment:
I would rewrite it just for beautifying:
```suggestion
        } else if static_pages != WasmPageNumber::zero() {
```

Reply:
ok